### PR TITLE
Make `checked` ops emit *unchecked* LLVM operations where feasible

### PR DESCRIPTION
**Pull request by scottmcm**
For things with easily pre-checked overflow conditions -- shifts and unsigned subtraction -- write the checked methods in such a way that we stop emitting wrapping versions of them.

For example, today <https://rust.godbolt.org/z/qM9YK8Txb> neither
```rust
a.checked_sub(b).unwrap()
```
nor
```rust
a.checked_sub(b).unwrap_unchecked()
```
actually optimizes to `sub nuw`.  After this PR they do.

cc #103299

### CODE REVIEW

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -1163,12 +1163,19 @@ macro_rules! int_impl {
...
         pub const fn checked_shl(self, rhs: u32) -> Option<Self> {
-            let (a, b) = self.overflowing_shl(rhs);
-            if unlikely!(b) { None } else { Some(a) }
+            // Not using overflowing_shl as that's a wrapping shift
+            if rhs < Self::BITS {
+                // SAFETY: just checked the RHS is in-range
+                Some(unsafe { self.unchecked_shl(rhs) })
+            } else {
+                None
+            }
```
Comment (workingjubilee) [inline library/core/src/num/int_macros.rs:1178]: It weirds me out slightly whenever I see the rhs here is a u32 instead of being generic over integers like the shl/shr operator itself is.

File: tests/mir-opt/pre-codegen/checked_ops.checked_shl.PreCodegen.after.panic-unwind.mir [hunk 2]
```diff
@@ -5,60 +5,33 @@ fn checked_shl(_1: u32, _2: u32) -> Option<u32> {
...
-            let mut _5: bool;
-            scope 4 (inlined core::num::<impl u32>::wrapping_shl) {
-                debug self => _1;
-                debug rhs => _2;
-                let mut _3: u32;
-                scope 5 (inlined core::num::<impl u32>::unchecked_shl) {
-                    debug self => _1;
-                    debug rhs => _3;
-                }
-            }
```
Comment (workingjubilee) [inline tests/mir-opt/pre-codegen/checked_ops.checked_shl.PreCodegen.after.panic-unwind.mir:28]: completely simplified away, nice.

File: tests/mir-opt/pre-codegen/checked_ops.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // skip-filecheck
-//@ compile-flags: -O -Zmir-opt-level=2 -Cdebuginfo=2
```
Comment (workingjubilee) [inline tests/mir-opt/pre-codegen/checked_ops.rs:2]: why this change?

File: tests/codegen/checked_math.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    // CHECK-DAG: %[[DIFF_U:.+]] = select i1 %[[IS_SOME]], i16 %[[DIFF_P]], i16 undef
+
+    // CHECK: %[[R0:.+]] = insertvalue { i16, i16 } poison, i16 %[[DISCR]], 0
+    // CHECK: %[[R1:.+]] = insertvalue { i16, i16 } %[[R0]], i16 %[[DIFF_U]], 1
+    // CHECK: ret { i16, i16 } %[[R1]]
+    a.checked_sub(b)
+}
+
+// Note that `shl` and `shr` in LLVM are already unchecked. So rather than
+// looking for no-wrap flags, we just need there to not be any masking.
```
[the remark this answers is not in the document]
Comment (workingjubilee) [reply tests/codegen/checked_math.rs:26]: looks like the same assembly emitted, tho'.

**Review by workingjubilee — COMMENTED** [review]
at first I thought "surely this doesn't need perf, how much could the compiler possibly stress over that?" but then I saw all the const eval machinery for implementing shl/shr does in fact flow through `checked_{shl,shr}`, thus gets used to implement it, thus gets squeezed, aiui, every time we compile a const in the shape of
```rust
const NAME: i32 = 1 << N;
```
right?

**Comment by workingjubilee** [comment]
would mostly expect the main difference to be from reduced LLVMIR if anything.

**Comment by DianQK** [comment]
Could this solve https://github.com/rust-lang/hashbrown/issues/509?

**Comment by nikic** [comment]
@scottmcm FWIW, for unsigned add overflow the recommended pattern is x + y < x, rather than uadd.with.overflow. We should probably give replacing that a try as well.

File: tests/mir-opt/pre-codegen/checked_ops.rs (hunk 3 above)
Comment (workingjubilee) [reply tests/mir-opt/pre-codegen/checked_ops.rs:2]: I don't mind it, either way, tbh, I just wanted to know the reasoning so that I know it wasn't Perfectly Random.

File: tests/codegen/checked_math.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (workingjubilee) [reply tests/codegen/checked_math.rs:26]: That's a good point! I tried an architecture that actually has a slightly more annoying shift instruction in this regard, PowerPC!
```asm
old_checked_sub:
        clrldi  5, 4, 32
        clrlwi  4, 4, 27
        addi 5, 5, -32
        srw 4, 3, 4
        rldicl 5, 5, 1, 63
        mr      3, 5
        blr
        .long   0
        .quad   0

new_checked_sub:
        clrldi  5, 4, 32
        slw 4, 3, 4
        addi 5, 5, -32
        rldicl 5, 5, 1, 63
        mr      3, 5
        blr
        .long   0
        .quad   0
```

One less instruction!

File: tests/mir-opt/pre-codegen/checked_ops.checked_shl.PreCodegen.after.panic-unwind.mir (hunk 2 above)
Comment (workingjubilee) [reply tests/mir-opt/pre-codegen/checked_ops.checked_shl.PreCodegen.after.panic-unwind.mir:28]: library-level opts are just being a very slow compiler. :^)

**Comment by workingjubilee** [comment]
Yeah, seems good to me!

@bors r+
